keps/sig-api-machinery/4933-cacher-label-index/README.md

```diff
@@ -0,0 +1,875 @@
+
+Considering that daemonset and spark applications coexist in a cluster, and they watch based on `spec.nodeName` and `applicationId` label respectively, we will support both usage at the same time.
+
+In reality, a single filter condition, such as nodeName or applicationId, can already accurately match cacheWatcher, this feature still considers supporting only one triggerIndex for a watch request. At the same time, since the following prerequisites guarantee that OR operations are not supported, filtering events with a single triggerIndex can ensure correctness
+- Label selectors and field selectors in a watch request are AND operations.
+- Multiple requirements in label selector are also AND operations.
+- Through a [`RequiresExactMatch`](https://github.com/kubernetes/kubernetes/blob/v1.31.2/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L574) check, a set-based label selector or field selector that supports OR operations on multiple values is considered not matching trigger indexes in cacher.
+
+### Commandline changes
+
+Considering that label names may be different in different scenarios, such as app label name in Flink and spark-app-selector label name in Spark, this feature will add a commandline parameter to the apiserver, allowing different label names to be configured as the trigger index of the apiserver cacher.
+
+```
+--index-labels. Index labels settings for some resources (pods, nodes, etc.), comma separated. The individual setting format: resource[.group]#labelName1;labelName2, where resource is lowercase plural (no version), group is omitted for resources of apiVersion v1 (the legacy core API) and included for others, and labelName is a string. It takes effect when watch-cache and feature gate CacherLabelIndex are enabled. In addition to built-in indexes such as pod.spec.nodeName, apiserver cacher will create additional indexes for these labels to speed up list and watch requests.
+
+# e.g. for spark
+--index-labels=pods#spark-app-selector
+# e.g. for spark and flink
+--index-labels=pods#spark-app-selector;app
+```
+
+### Changes of register trigger index and cache index to resource storage
+
+In order to passing the commandline label indexes parameters into resource storage and cacher, we'll add a `IndexLabels []string` field into `Config` struct in file staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go
+
+We will parse the commandline label indexes parameters in `Store.CompleteWithOptions` and register them into `Indexers` and `TriggerFunc` of `StoreOptions`.
+
+### Major struct and field changes in cacher
+
+Currently `indexedTrigger *indexedTriggerFunc` is a single function field in cacher to support field index. It is assigned to the `spec.nodeName` function for pod resource.
+
+We will update it to `indexedTriggers map[string]storage.IndexerFunc` to support a single field index and multiple label indexes. Multiple field indexes is not supported as before and is beyond the scope of this KEP.
```

Comment:
If we want to support multiple label indexes, why can't we support multiple field indexes too?
With the changes you're proposing - won't we effectively get it for free?

Reply:
Yes, it can be supported, but I don't need this feature for now. If you think this is required, I can supported along this feature.

Reply:
If we want to touch it, we should support both - this won't introduce visible additional overhead, I think.